### PR TITLE
Feature/sc 25483

### PR DESCRIPTION
**Pull request by KuphJr**
## Changes

- Added Overrider class to ea-bootstrap to handle symbol-to-symbol and symbol-to-id overrides
- Implemented the Overrider class for the CoinGecko adapter

## Quality Assurance

- [x] Ran `yarn changeset` if adapter source code was changed
- [x] The branch naming follows git flow (`feature/x`, `chore/x`, `release/x`, `hotfix/x`, `fix/x`) or is created from Clubhouse/Shortcut
- [x] This is related to a maximum of one Clubhouse/Shortcut story or GitHub issue
- [x] Types are safe (avoid TypeScript/TSLint features like any and disable, instead use more specific types)

### CODE REVIEW

File: packages/sources/coingecko/src/endpoint/crypto.ts [hunk 1]
```diff
@@ -54,6 +72,14 @@ export const inputParameters: InputParameters = {
     description: 'The symbol of the currency to convert to',
     required: true,
   },
+  overrides: {
```
Comment (justinkaseman) [inline packages/sources/coingecko/src/endpoint/crypto.ts:75]: These are merged in from the base of https://github.com/smartcontractkit/external-adapters-js/blob/develop/packages/core/bootstrap/src/lib/modules/selector.ts#L15

File: packages/sources/coingecko/src/endpoint/crypto.ts [hunk 2]
```diff
@@ -36,7 +45,16 @@ export const endpointResultPaths: {
 }
 
 export const description =
-  '**NOTE: the `price` endpoint is temporarily still supported, however, is being deprecated. Please use the `crypto` endpoint instead.**'
+  '**NOTE: the `price` endpoint is temporarily still supported, however, is being deprecated. Please use the `crypto` endpoint instead.**\n' +
+  'Overrides can be provided as input parameters, as specified below, or be hardcoded into the `symbolToSymbolOverrides.json` or `symbolToIdOverrides.json` files in the `./src/config folder`.  Overrides specified as input parameters take precedence.  The order of operations for performing overrides is as follows: \n' +
```
Comment (justinkaseman) [inline packages/sources/coingecko/src/endpoint/crypto.ts:49]: This seems pretty overkill here. A README in the Override class would be a better place for this.

File: packages/sources/coingecko/src/endpoint/crypto.ts [hunk 3]
```diff
@@ -101,20 +139,38 @@ const handleBatchedRequest = (
...
+  const validator = new Validator(
+    request,
+    inputParameters,
+    {},
+    {
+      overrides,
+      symbolToIdOverrides,
+    },
+  )
+  const endpoint = validator.validated.data.endpoint ?? DEFAULT_ENDPOINT
```
Comment (justinkaseman) [inline packages/sources/coingecko/src/endpoint/crypto.ts:151]: This should be already filled from the `Selector`. Are you seeing scenarios where it isn't?

File: packages/core/bootstrap/test/unit/validator.test.ts [hunk 4]
```diff
@@ -548,6 +620,8 @@ describe('Validator', () => {
       expect(base).toBe('btc')
     })
 
+    /* The following tests have been removed since symbol to symbol and symbol to id overriders
```
Comment (justinkaseman) [inline packages/core/bootstrap/test/unit/validator.test.ts:623]: Something like this you should fully remove the code for and then you can leave this comment on the diff here.

File: packages/core/bootstrap/src/lib/modules/validator.ts [hunk 5]
```diff
@@ -161,6 +203,7 @@ export class Validator {
     return pairs[0].includes[0]
   }
 
+  // !!! function no longer used
```
Comment (justinkaseman) [inline packages/core/bootstrap/src/lib/modules/validator.ts:206]: Should be removed then.

File: packages/core/bootstrap/src/lib/modules/validator.ts [hunk 6]
```diff
@@ -119,6 +160,7 @@ export class Validator {
     }
   }
 
+  // !!!! Function no longer used
```
Comment (justinkaseman) [inline packages/core/bootstrap/src/lib/modules/validator.ts:163]: Should be removed then.

File: package.json [hunk 7]
```diff
@@ -55,6 +55,7 @@
     "@types/node": "16.11.19",
     "husky": "7.0.4",
     "lint-staged": "11.2.6",
+    "nock": "^13.2.4",
```
Comment (justinkaseman) [inline package.json:58]: Why is this going to root? Should be removed.

**Comment by tateexon** [comment]
Soak test for coingecko failed: `     ✗ Group-0-coingecko-SHIB / ETH returned 200
      ↳  0% — ✓ 0 / ✗ 620
     ✗ Group-0-coingecko-SWAP / ETH returned 200
      ↳  0% — ✓ 0 / ✗ 620
     ✗ Group-0-coingecko-ATOM / USD returned 200
      ↳  20% — ✓ 127 / ✗ 493
     ✗ Group-0-coingecko-KNC / USD returned 200
      ↳  21% — ✓ 131 / ✗ 489
     ✗ Group-0-coingecko-BETA / ETH returned 200
      ↳  0% — ✓ 0 / ✗ 620
     ✗ Group-0-coingecko-WOM / ETH returned 200
      ↳  28% — ✓ 178 / ✗ 442
     ✗ Group-0-coingecko-BAND / USD returned 200
      ↳  23% — ✓ 143 / ✗ 477
     ✗ Group-0-coingecko-CRO / USD returned 200
      ↳  26% — ✓ 164 / ✗ 456
     ✗ Group-0-coingecko-OCEAN / ETH returned 200
      ↳  24% — ✓ 150 / ✗ 470
     ✗ Group-0-coingecko-FOX / USD returned 200
      ↳  0% — ✓ 0 / ✗ 620
     ✗ Group-0-coingecko-ADX / USD returned 200
      ↳  0% — ✓ 0 / ✗ 620
     ✗ Group-0-coingecko-1INCH / ETH returned 200
      ↳  0% — ✓ 0 / ✗ 620
     ✗ Group-0-coingecko-PUNDIX / USD returned 200
      ↳  26% — ✓ 165 / ✗ 455
     ✗ Group-0-coingecko-MANA / USD returned 200
      ↳  0% — ✓ 0 / ✗ 620
     ✗ Group-0-coingecko-TRX / USD returned 200
      ↳  0% — ✓ 0 / ✗ 620
     ✗ Group-0-coingecko-ZCN / USD returned 200
      ↳  24% — ✓ 150 / ✗ 470
     ✗ Group-0-coingecko-CRO / ETH returned 200
      ↳  25% — ✓ 161 / ✗ 459
     ✗ Group-0-coingecko-OMG / USD returned 200
      ↳  22% — ✓ 138 / ✗ 482
     ✗ Group-0-coingecko-FRAX / ETH returned 200
      ↳  0% — ✓ 0 / ✗ 620
     ✗ Group-0-coingecko-BTC / ETH returned 200
      ↳  25% — ✓ 156 / ✗ 464
     ✗ Group-0-coingecko-BAL / ETH returned 200
      ↳  25% — ✓ 155 / ✗ 465
     ✗ Group-0-coingecko-COMP / USD returned 200
      ↳  27% — ✓ 172 / ✗ 448
     ✗ Group-0-coingecko-BORING / USD returned 200
      ↳  27% — ✓ 169 / ✗ 451
     ✗ Group-0-coingecko-NU / ETH returned 200
      ↳  25% — ✓ 156 / ✗ 464
     ✗ Group-0-coingecko-ACH / USD returned 200
      ↳  23% — ✓ 148 / ✗ 472
     ✗ Group-0-coingecko-ZEC / USD returned 200
      ↳  29% — ✓ 182 / ✗ 438
     ✗ Group-0-coingecko-ONT / USD returned 200
      ↳  0% — ✓ 0 / ✗ 620
     ✗ Group-0-coingecko-DOGE / USD returned 200
      ↳  0% — ✓ 0 / ✗ 620
     ✗ Group-0-coingecko-CREAM / ETH returned 200
      ↳  0% — ✓ 0 / ✗ 620
     ✗ Group-0-coingecko-FEI / USD returned 200
      ↳  0% — ✓ 0 / ✗ 620
     ✗ Group-0-coingecko-ADA / USD returned 200
      ↳  0% — ✓ 0 / ✗ 620
     ✗ Group-0-coingecko-ANKR / ETH returned 200
      ↳  0% — ✓ 0 / ✗ 620
     ✗ Group-0-coingecko-AAVE / USD returned 200
      ↳  0% — ✓ 0 / ✗ 620
     ✗ Group-0-coingecko-STAKE / ETH returned 200
      ↳  23% — ✓ 146 / ✗ 474
     ✗ Group-0-coingecko-ZRX / ETH returned 200
      ↳  0% — ✓ 0 / ✗ 620
     ✗ Group-0-coingecko-MKR / USD returned 200
      ↳  24% — ✓ 152 / ✗ 468
     ✗ Group-0-coingecko-BCH / USD returned 200
      ↳  27% — ✓ 168 / ✗ 452
     ✗ Group-0-coingecko-EURT / USD returned 200
      ↳  0% — ✓ 0 / ✗ 620
     ✗ Group-0-coingecko-SXP / USD returned 200
      ↳  0% — ✓ 0 / ✗ 620
     ✗ Group-0-coingecko-AMPL / ETH returned 200
      ↳  0% — ✓ 0 / ✗ 620
     ✗ Group-0-coingecko-ETH / USD returned 200
      ↳  0% — ✓ 0 / ✗ 620
     ✗ Group-0-coingecko-CEL / ETH returned 200
      ↳  0% — ✓ 0 / ✗ 620
     ✗ Group-0-coingecko-SRM / ETH returned 200
      ↳  30% — ✓ 187 / ✗ 433
     ✗ Group-0-coingecko-LUNA / ETH returned 200
      ↳  0% — ✓ 0 / ✗ 620
     ✗ Group-0-coingecko-C98 / USD returned 200
      ↳  0% — ✓ 0 / ✗ 620
     ✗ Group-0-coingecko-BNT / ETH returned 200
      ↳  0% — ✓ 0 / ✗ 620
     ✗ Group-0-coingecko-HEGIC / ETH returned 200
      ↳  21% — ✓ 136 / ✗ 484
     ✗ Group-0-coingecko-IOST / USD returned 200
      ↳  24% — ✓ 154 / ✗ 466
     ✗ Group-0-coingecko-BUSD / USD returned 200
      ↳  22% — ✓ 141 / ✗ 479
     ✗ Group-0-coingecko-ZRX / USD returned 200
      ↳  0% — ✓ 0 / ✗ 620
     ✗ Group-0-coingecko-AUTO / USD returned 200
      ↳  0% — ✓ 0 / ✗ 620
     ✗ Group-0-coingecko-WNXM / ETH returned 200
      ↳  20% — ✓ 128 / ✗ 492
     ✗ Group-0-coingecko-SNX / ETH returned 200
      ↳  0% — ✓ 0 / ✗ 620
     ✗ Group-0-coingecko-XTZ / USD returned 200
      ↳  25% — ✓ 156 / ✗ 464
     ✗ Group-0-coingecko-MKR / ETH returned 200
      ↳  21% — ✓ 136 / ✗ 484
     ✗ Group-0-coingecko-DAI / USD returned 200
      ↳  0% — ✓ 0 / ✗ 620
     ✗ Group-0-coingecko-STMX / USD returned 200
      ↳  26% — ✓ 165 / ✗ 455
     ✗ Group-0-coingecko-DODO / USD returned 200
      ↳  24% — ✓ 151 / ✗ 469
     ✗ Group-0-coingecko-SOL / USD returned 200
      ↳  0% — ✓ 0 / ✗ 620
     ✗ Group-0-coingecko-RGT / ETH returned 200
      ↳  0% — ✓ 0 / ✗ 620
     ✗ Group-0-coingecko-AVAX / USD returned 200
      ↳  0% — ✓ 0 / ✗ 620
     ✗ Group-0-coingecko-OXT / USD returned 200
      ↳  30% — ✓ 191 / ✗ 429
     ✗ Group-0-coingecko-USDT / USD returned 200
      ↳  0% — ✓ 0 / ✗ 620
     ✗ Group-0-coingecko-USDC / ETH returned 200
      ↳  0% — ✓ 0 / ✗ 620
     ✗ Group-0-coingecko-USDT / ETH returned 200
      ↳  0% — ✓ 0 / ✗ 620

     checks.........................: 11.33% ✓ 16452    ✗ 128628
     data_received..................: 62 MB  1.4 kB/s
     data_sent......................: 19 MB  448 B/s
     errors.........................: 88.66% ✓ 128628   ✗ 16452 
     http_req_blocked...............: avg=5.93µs  min=86ns     med=272ns    max=443.64ms p(90)=640ns    p(95)=693ns   
     http_req_connecting............: avg=65ns    min=0s       med=0s       max=1.13ms   p(90)=0s       p(95)=0s      
   ✗ http_req_duration..............: avg=1.73s   min=1.09ms   med=2.5s     max=1m0s     p(90)=2.52s    p(95)=2.61s   
       { expected_response:true }...: avg=1.15s   min=5.3ms    med=159.87ms max=6.55s    p(90)=2.75s    p(95)=2.86s   
   ✗ http_req_failed................: 88.65% ✓ 128822   ✗ 16487 
     http_req_receiving.............: avg=71.36µs min=0s       med=61.73µs  max=36.95ms  p(90)=102.05µs p(95)=116.14µs
     http_req_sending...............: avg=74.23µs min=12.77µs  med=61.01µs  max=13.61ms  p(90)=124.02µs p(95)=139.82µs
     http_req_tls_handshaking.......: avg=1.61µs  min=0s       med=0s       max=190.2ms  p(90)=0s       p(95)=0s      
     http_req_waiting...............: avg=1.73s   min=988.73µs med=2.5s     max=1m0s     p(90)=2.52s    p(95)=2.61s   
     http_reqs......................: 145309 3.361299/s
     iteration_duration.............: avg=1m9s    min=6.72s    med=1m9s     max=1m44s    p(90)=1m12s    p(95)=1m14s   
     iterations.....................: 620    0.014342/s
     vus............................: 1      min=1      max=1   
     vus_max........................: 1      min=1      max=1   `

File: packages/core/bootstrap/test/unit/validator.test.ts (hunk 4 above)
Comment (KuphJr) [reply packages/core/bootstrap/test/unit/validator.test.ts:623]: I think there may be other adapters still relying on this old code, hence why it has not been removed.  This code will be removed once the overrider is implemented for the remaining adapters.

File: packages/core/bootstrap/src/lib/modules/validator.ts (hunk 5 above)
Comment (KuphJr) [reply packages/core/bootstrap/src/lib/modules/validator.ts:206]: Added JSdoc '@deprecated' comment

File: packages/core/bootstrap/src/lib/modules/validator.ts (hunk 6 above)
Comment (KuphJr) [reply packages/core/bootstrap/src/lib/modules/validator.ts:163]: Added JSdoc '@deprecated' comment

File: packages/sources/coingecko/src/endpoint/crypto.ts (hunk 2 above)
Comment (KuphJr) [reply packages/sources/coingecko/src/endpoint/crypto.ts:49]: I removed the description here and moved it to the bootstrap/lib/modules README file.

File: packages/sources/coingecko/src/endpoint/crypto.ts (hunk 1 above)
Comment (KuphJr) [reply packages/sources/coingecko/src/endpoint/crypto.ts:75]: I added symbolToIdOverrides to the baseInputParameters object in lib/modules/selector.ts

File: packages/sources/coingecko/src/endpoint/crypto.ts (hunk 3 above)
Comment (KuphJr) [reply packages/sources/coingecko/src/endpoint/crypto.ts:151]: I am not quite sure what you mean here.  The only change from the original code on line 104 was to add symbolToIdOverrides that are specified in the adapter's symbolToIdOverrides.json file

**Review by KuphJr — COMMENTED** [review]
I addressed all of your comments from your review & all tests are passing

**Pull request closed without merging**